Add CSSTransitionGroup info

joshwcomeau opened the pull request:
:wave: hi folks!

As discussed [on twitter](https://twitter.com/_philpl/status/881249767669583873), adding some documentation for integrating styled-components with CSSTransitionGroup.

Added this to the FAQ, but I'm not sure that that's the best place for it. Lemme know!

![screen shot 2017-07-08 at 11 46 36 am](https://user-images.githubusercontent.com/6692932/27986863-26f7e0fe-63d3-11e7-964d-428a5e8f6099.png)

joshwcomeau commented on components/faqs/css-transition-group.js:
```diff
@@ -0,0 +1,63 @@
+`).trim()
+
+const Nesting = () => (
+  <SectionLayout title="Integrating with CSSTransitionGroup">
```
So, the other FAQ items have question-titles (eg. "Can I refer to other components?"). I'm breaking with convention here :x

The reason is that the menu items aren't in question-format, and the menu item needs to match the section header for the hash anchors to work. I've opened an issue (https://github.com/styled-components/styled-components-website/issues/53), but in the meantime I've opted to make the link work, over following the heading convention.

philpl replied:
I'll fix this soon :) Thanks for noticing!

joshwcomeau commented on components/faqs/css-transition-group.js:
```diff
@@ -0,0 +1,63 @@
+import {CSSTransitionGroup} from 'react-transition-group'
+
+// Define constants for all our class names and timeouts.
+// This is optional, but recommended.
```
Feel free to disagree with me on this recommendation!

philpl commented on components/faqs/css-transition-group.js:
```diff
@@ -0,0 +1,63 @@
+
+const Nesting = () => (
+  <SectionLayout title="Integrating with CSSTransitionGroup">
+    <p><a href="https://github.com/reactjs/react-transition-group"><Code>react-transition-group</Code></a> is a popular React package for animating components as they enter or leave.</p>
```
Can you use the `Link` component here please? You can see how it's used in some other sections

philpl commented on components/faqs/css-transition-group.js:
```diff
@@ -0,0 +1,63 @@
+const Nesting = () => (
+  <SectionLayout title="Integrating with CSSTransitionGroup">
+    <p><a href="https://github.com/reactjs/react-transition-group"><Code>react-transition-group</Code></a> is a popular React package for animating components as they enter or leave.</p>
+    <p><Code>CSSTransitionGroup</Code>, the high-level animation component exported by react-transition-group, works by applying class names to the components when they're transitioning; they leave the exact animation implementation up to you.</p>
```
`react-transition-group` needs to be wrapped or not wrapped above for consistency ;)

joshwcomeau replied:
Good catch!

mxstbr commented:
Damn now we got them conflicts going 😕 Sorry, mind resolving them so we can finally land this?

tkvw commented:
Just a heads-up, CSSTransitionGroup is removed from V2 of react-transition-group. Maybe it's best to include an example with the latest version? 
@see: https://github.com/reactjs/react-transition-group/blob/master/Migration.md

joshwcomeau commented:
Whoops! Sorry for the delay, totally fell off my radar.

Some stuff:

- While fixing this I noticed a bug with a sibling faq component. I addressed it in #66, and it should be merged first (this branch won't work without it)

- For `react-transition-group` v2, I added a note to clarify which version the docs are written for. V1 is still under active maintenance, and I suspect the overwhelming majority of users haven't upgraded yet, so I think it makes sense to keep our docs for v1, for now? Truthfully part of this is that I don't have the time to upgrade my other project and see if these instructions are still good.

philpl commented:
@joshwcomeau you can ignore the CI, as I haven't added a `.travis.yml` file yet, which is simply due to a lack of time on my end

davelab commented:
I'm trying using this example with the new react-transition-group's API v2 and also v1 as shown in this example. But it seems that doesn't work properly. the class names are never applied to the children.

`import React from 'react';
import styled, { css } from 'styled-components';
import CSSTransition from 'react-transition-group/CSSTransition';

const animationClassName = 'fade';
const enter = 'fade-enter';
const enterActive = 'fade-enter-active';
const exit = 'fade-exit';
const exitActive = 'fade-exit-active';
const enterTimeout = 1000;
const exitTimeout = 1000;

const style = css`
  .${enter} {
    opacity: 0.01;
  }
  .${enterActive} {
    opacity: 1;
    transition: opacity ${enterTimeout}ms ease-in;
  }
  .${exit} {
    opacity: 1;
  }
  .${exitActive} {
    opacity: 0.01;
    transition: opacity ${exitTimeout}ms ease-in;
  }
`;
export default styled(CSSTransition).attrs({
  classNames: animationClassName,
  timeout: enterTimeout
})`
  ${style}
`;
`
and the output is this:
![schermata 2017-08-17 alle 15 13 50](https://user-images.githubusercontent.com/719197/29414090-ea52f250-835e-11e7-8201-26af71b904a1.png)
![schermata 2017-08-17 alle 15 14 20](https://user-images.githubusercontent.com/719197/29414091-ea560d32-835e-11e7-8f94-8fd9250983b8.png)

Can someone review it? thanks

jenovs commented:
@davelab If you're using TransitionGroup as a wrapper I'd say that classes aren't applied because you aren't providing a unique key to your styled CSSTransition component.

Also you should add `&`s in your styles:
```
  &.${enter} {
    opacity: 0.01;
  ...
```

https://stackblitz.com/edit/react-629q2j?file=index.js

philpl commented:
We've changed how to write new content (also to ease contributions for translations in the future) and unfortunately this will have a lot of conflicts now.

Also, it seems that there are some conflicting information here. And as it's not essential information for the usage of styled-components how to use react-transition-group I think we should close this for now. Feel free though to update it, if you still would like this to go in. Sorry! 😞 

Ping me if you need some help updating it. Happy to help, or take it over from here!

mxstbr commented:
Awww, I really wanted to have this! :cry:

philpl commented:
@mxstbr it’s not hard to port, if you’d like 😜 but I personally am not comfortable putting it in with the above comments not being addressed fully

joshwcomeau commented:
Fair enough! Yeah, truthfully I don't really wanna shoulder the burden of maintaining this section of the docs. I really like styled-components, but at Khan Academy we use Aphrodite, and most of my side projects use something else.

Also - sorry to the folks that commented asking for help. The example in the doc works in my project, but I don't have the depth of knowledge to speculate as to why it isn't working for y'all :( hope it worked out in the end!
